haunt-roll-fail/arcs/game.scala

```diff
@@ -1,2728 +1,2744 @@
+
+            case ExecuteMainAction(f, x, then) =>
+                implicit val convert = (u : Figure, selected : Boolean) => {
+                    val prefix = selected.not.??(u.faction.?./(_.short.toLowerCase + "-").|(""))
+                    val suffix = selected.??("-empty")
+
+                    Image(prefix + "agent" + suffix, styles.qship)
+                }
+
+                XXSelectObjectsAction(f, f.captives)
+                    .withGroup("Execute captives", x)
+                    .withThen(l => ExecuteAction(f, l, x, then))(l => ("Execute", l))
+                    .withExtras(CancelAction)
+
+            case ExecuteAction(f, l, x, then) =>
+                f.pay(x)
+
+                f.captives --> l --> f.trophies
+
+                f.log("executed", l, x)
+
+                then
+
+            case AbductMainAction(f, l, x, then) =>
+                Ask(f).group("Abduct".hl, "Agents", x)
+                    .each(market)(c => AbductAction(f, c, x, then).as(c).!(l.has(c).not))
+                    .cancel
+
+            case AbductAction(f, c, x, then) =>
+                f.pay(x)
+
+                val l = Influence(c).%(_.faction != f)
+
+                l --> f.captives
+
+                f.log("abducted", l, x)
+
+                then
+
+            case FarseersMainAction(f, e, then) =>
+                YYSelectObjectsAction(f, e.hand)
+                    .withGroup(Farseers, "look at", e, "cards")
+                    .withThen(d => FarseersAction(f, e, d, then))(d => ("Take", d, "from", e))("Take a card from", e)
+                    .withExtras(then.as("Skip"))
+
+            case FarseersAction(f, e, d, then) =>
+                e.hand --> d --> f.hand
+
+                f.log("took a card from", e, "with", Farseers)
+
+                YYSelectObjectsAction(f, f.hand)
+                    .withGroup(Farseers, "give back", e, "a card")
+                    .withThen(d => FarseersBackAction(f, e, d, then))(d => ("Give", d, "to", e))("Give a card back to", e)
+                    .withExtras(NoHand)
+
+            case FarseersBackAction(f, e, d, then) =>
+                f.hand --> d --> e.hand
+
+                f.log("gave back a card to", e)
+
+                then
+
+            case FarseersRedrawMainAction(f, then) =>
+                XXSelectObjectsAction(f, f.hand)
+                    .withGroup(Farseers, "discard cards to redraw")
+                    .withThen(l => FarseersRedrawAction(f, l, then))(l => ("Discard to draw", (l.num + 1).cards))
+                    .withExtras(NoHand, FarseersRedrawAction(f, $, then).as("Discard", Farseers, "to draw", 1.cards), CancelAction)
+
+            case FarseersRedrawAction(f, l, then) =>
+                val d = deck.num
+
+                f.hand --> l --> deck
+
+                if (l.num + 1 > d && disdeck.any) {
+                    Shuffle[DeckCard](deck.$ ++ disdeck.$, ShuffleDeckCardsAction(_, FarseersRedrawAction(f, l, then)))
```

Comment:
This won't work as written as Shuffle is a value, not a function.
A value that would be ignored unless it is returned to the engine on the last line of a _case block_.

And it's a value cause we have to delegate any randomness to the engine so that when we replay actions we always get the same game state - if there were any dice rolls or shuffles, they have to be recorded as separate actions.

Something like this might work:
```
case FarseersRedrawAction(f, l, then) if l.num + 1 > deck.num && disdeck.any =>
    Shuffle[DeckCard](deck.$ ++ disdeck.$, ShuffleDeckCardsAction(_, FarseersRedrawAction(f, l, then)))

case FarseersRedrawAction(f, l, then) if l.num + 1 > deck.num && f.hand.any =>
    Shuffle[DeckCard](deck.$ ++ f.hand.$, ShuffleDeckCardsAction(_, FarseersRedrawAction(f, l, then)))

case FarseersRedrawAction(f, l, then) =>
    l.foreach { d =>
        if (f.hand.has(d))
            d --> deck
    }
...
```
Or
```
if (l.num + 1 > d && disdeck.any) {
    Shuffle[DeckCard](deck.$ ++ disdeck.$, ShuffleDeckCardsAction(_, FarseersRedrawAction(f, l, then)))
}
else { // !!
    ...
}
```

Reply:
Changed to use if/else so the last action is the happy path.